internal/cf/client.go

```diff
@@ -92,71 +98,101 @@ func newSpaceClient(spaceGuid string, url string, username string, password stri
-	orgClientCache   = make(map[clientIdentifier]*organizationClient)
-	cacheMutex       = &sync.Mutex{}
-)
-
 func NewOrganizationClient(organizationName string, url string, username string, password string) (facade.OrganizationClient, error) {
 	cacheMutex.Lock()
 	defer cacheMutex.Unlock()
-	identifier := clientIdentifier{URL: url, Username: username}
-	client, cached := orgClientCache[identifier]
-	var err error
-	if !cached {
+
+	// look up CF client in cache
+	identifier := clientIdentifier{url: url, username: username}
+	cacheEntry, isInCache := clientCache[identifier]
+
+	var err error = nil
+	var client *organizationClient = nil
+	if isInCache {
+		// re-use CF client and wrap it as organizationClient
+		client = &organizationClient{organizationName: organizationName, client: cacheEntry.client}
+		if cacheEntry.password != password {
+			// password was rotated => delete client from cache and create a new one below
+			delete(clientCache, identifier)
+			isInCache = false
+		}
+	}
+
+	if !isInCache {
+		// create new CF client and wrap it as organizationClient
 		client, err = newOrganizationClient(organizationName, url, username, password)
 		if err == nil {
-			orgClientCache[identifier] = client
-		}
-	} else {
-		// If the password has changed since we cached the client, we want to update it to the new one
-		if client.password != password {
-			client.password = password
+			// add CF client to cache
+			clientCache[identifier] = &clientCacheEntry{url: url, username: username, password: password, client: client.client}
 		}
 	}
+
 	return client, err
 }
 
 func NewSpaceClient(spaceGuid string, url string, username string, password string) (facade.SpaceClient, error) {
 	cacheMutex.Lock()
 	defer cacheMutex.Unlock()
-	identifier := clientIdentifier{URL: url, Username: username}
-	client, cached := spaceClientCache[identifier]
-	var err error
-	if !cached {
+
+	// look up CF client in cache
+	identifier := clientIdentifier{url: url, username: username}
+	cacheEntry, isInCache := clientCache[identifier]
+
+	var err error = nil
+	var client *spaceClient = nil
+	if isInCache {
+		// re-use CF client from cache and wrap it as spaceClient
+		client = &spaceClient{spaceGuid: spaceGuid, client: cacheEntry.client}
+		if cacheEntry.password != password {
+			// password was rotated => delete client from cache and create a new one below
+			delete(clientCache, identifier)
+			isInCache = false
+		}
+	}
+
+	if !isInCache {
+		// create new CF client and wrap it as spaceClient
 		client, err = newSpaceClient(spaceGuid, url, username, password)
 		if err == nil {
-			spaceClientCache[identifier] = client
-		}
-	} else {
-		// If the password has changed since we cached the client, we want to update it to the new one
-		if client.password != password {
-			client.password = password
+			// add CF client to cache
+			clientCache[identifier] = &clientCacheEntry{url: url, username: username, password: password, client: client.client}
```

Comment:
For now I think we should merge this, but later think about factoring out this part.